contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,437 @@
+        // This parameter sets a boundary on the pool's swap price. It defines the
+        // worst acceptable price before the transaction reverts to allow for partial
+        // swaps, hence the value here is zero as we do not want to accept partial
+        // swaps.
+        ISwapRouter.ExactInputSingleParams memory params =
+            ISwapRouter.ExactInputSingleParams({
+                tokenIn: tokenIn,
+                tokenOut: tokenOut,
+                fee: fee,
+                recipient: recipient,
+                deadline: block.timestamp,
+                amountIn: amountIn,
+                amountOutMinimum: amountOutMinimum,
+                sqrtPriceLimitX96: 0
+            });
+
+        // execute the swap
+        amountOut = swapRouter.exactInputSingle(params);
+    }
+
+    /**
+     * @notice Collects payment funds and performs currency swap for cross-currency loan repayment.
+     *         Any funds that are not covered by closing out the old loan must be covered by
+     *         the borrower.  Any excess funds are sent to the borrower.
+     *
+     * @param borrower               The address of the borrower.
+     *                               executing the repayment and swap.
+     * @param lender                 The address of the new lender.
+     * @param swapParams             The parameters for the currency swap.
+     * @param newLoanTerms           The terms of the new loan.
+     * @param oldLoanData            The loan data of the original loan.
+     * @param oldLoanId              The ID of the original loan to be rolled over.
+     *
+     * @return repayAmount           The amount to be repaid to the old loan.
+     * @return feeSnapshot           A snapshot of current lending fees.
+     */
+    function _processSettlement(
+        address borrower,
+        address lender,
+        OriginationLibrary.SwapParameters memory swapParams,
+        LoanLibrary.LoanTerms memory newLoanTerms,
+        LoanLibrary.LoanData memory oldLoanData,
+        uint256 oldLoanId
+    ) internal returns (uint256 repayAmount, LoanLibrary.FeeSnapshot memory feeSnapshot){
+        // pull funds from the new lender
+        IERC20(newLoanTerms.payableCurrency).safeTransferFrom(
+            lender,
+            address(this),
+            newLoanTerms.principal
+        );
+
+        // swap the new lender payment from new currency to the original currency
+        // the swap will revert if swappedAmount < minAmountOut
+        uint256 swappedAmount = _swapExactInputSingle(
+            newLoanTerms.payableCurrency,
+            oldLoanData.terms.payableCurrency,
+            newLoanTerms.principal,
+            swapParams.minAmountOut,
+            swapParams.poolFeeTier,
+            address(this)
+        );
+
+        address oldLender = ILoanCore(loanCore).lenderNote().ownerOf(oldLoanId);
+
+        // get fee snapshot from fee controller
+        (feeSnapshot) = feeController.getFeeSnapshot();
```

Comment:
spaghetti code, this does not need to be here, can be moved to the function it is used in.

Reply:
Additionally re-taking the fee snapshot goes against how regular rollovers work. In a regular rollover, the new loan inherits the fee snapshot taken for the old loan. This was reported in one of our audits to do it this way